### ISSUE TITLE
TST: errors testing with CUDA

### ISSUE THREAD

**Issue by tylerjereddy** [kokkos/pykokkos#1]
I tried to follow the steps in the `README` to get setup, though one point of deviation is that I had to allow `pykokkos-base` to build its own vendored kokkos (it really was pretty picky about CMake-related problems with the pre-built Kokkos, not to mention that the docs here require building two separate Kokkos (OMP/CUDA) while the pykokkos-base docs don't seem to integrate so well with that).

When I run i.e., `CUDA_VISIBLE_DEVICES=0 pytest tests`

I get errors like this:

```
________________________________________________________________________________________________________________________ ERROR collecting tests/test_views.py _________________________________________________________________________________________________________________________
tests/test_views.py:4: in <module>
    import pykokkos as pk
pykokkos/__init__.py:14: in <module>
    initialize()
pykokkos/kokkos_manager/__init__.py:91: in initialize
    kokkos.initialize()
E   RuntimeError: cudaGetDeviceCount(&m_cudaDevCount) error( cudaErrorInitializationError): initialization error /home/tyler/github_projects/pykokkos-base/external/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:224
E   Traceback functionality not available

```

If I tell it to use a GPU that does not exist, I can at least understand the error message a little better:

`CUDA_VISIBLE_DEVICES=1 pytest tests`

```
________________________________________________________________________________________________________________________ ERROR collecting tests/test_views.py _________________________________________________________________________________________________________________________
tests/test_views.py:4: in <module>
    import pykokkos as pk
pykokkos/__init__.py:14: in <module>
    initialize()
pykokkos/kokkos_manager/__init__.py:91: in initialize
    kokkos.initialize()
E   RuntimeError: cudaGetDeviceCount(&m_cudaDevCount) error( cudaErrorNoDevice): no CUDA-capable device is detected /home/tyler/github_projects/pykokkos-base/external/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:224
E   Traceback functionality not available

```

Do I need a debug build to get the actual traceback? What do you suggest? Rebuilding `pykokkos-base` is super expensive for me: almost 2 hours as noted in https://github.com/kokkos/pykokkos-base/issues/31.

**Comment by jrmadsen** [comment]
What issues exist with pykokkos-base and an external Kokkos? The “vendored” Kokkos is just a clone of the repo. Does pykokkos also link to the Kokkos libs? You definitely do not want to have pykokkos and pykokkos-base loading different installs of Kokkos. Something like that could cause Kokkos initialization issues.

**Comment by tylerjereddy** [comment]
I think the docs are pretty clear that it can be a bit of a mess with `python setup.py install .. ` or `pip install` + CMake.

I had to:

- delete the  `kokkos_compilation` line in cmake/Modules/KokkosPythonKokkos.cmake` because of an error
- `export Kokkos_DIR=/home/tyler/github_projects/Kokkos/kokkos/build_openmp_cuda`
- then I run into stuff like: `ninja: error: 'Kokkos::kokkoscore-NOTFOUND', needed by 'kokkos/libpykokkos.cpython-38-x86_64-linux-gnu.so'`
- then I start messing around with `CMAKE_PREFIX_PATH` to help CMake out a bit...

Basically, `pykokkos-base` has a hard time finding the the stuff I built following the instructions in this repo. In fact, this repo only mentions `pykokkos-base` with a link to its repo, and actually nothing about it in the install process proper.

**Comment by NaderAlAwar** [comment]
Hi @tylerjereddy, I just pushed some changes that should fix this. Previously, pykokkos would use a pre-compiled version of pykokkos-base (committed to the repo) to avoid having to re-compile pykokkos-base. However that will cause issues similar to the one you're facing. The changes I pushed should get pykokkos to use the pykokkos-base you installed to your environment, which should fix the issues you're having.

pykokkos will link to the kokkos installation at the path specified by the environment variables `PK_KOKKOS_LIB_PATH_CUDA` and `PK_KOKKOS_LIB_PATH_OMP`. You should set these variables to the versions of Kokkos that were installed during the setup of pykokkos-base. So for example, in my case these would be located at `$PATH_TO_ENV/pykbase_final_pip/lib/python3.9/site-packages/pykokkos_base-0.0.7-py3.9-linux-x86_64.egg/lib/` (the last directory could also be `lib64/`)

The pykokkos README is slightly out of date, we're still working on fixing the installation process, but I will try to include the updated build instructions in the README.  

Let me know if you're still having issues

**Comment by tylerjereddy** [comment]
@NaderAlAwar I'm on the latest `develop` of `pykokkos`, let me know if I need another branch.

**Comment by NaderAlAwar** [comment]
No that's the one that you should be on. Based on your first comment, I'm guessing a simple script like this would reproduce the error?

```python
import kokkos

kokkos.initialize()
```

 I'm not sure how you would get a stack trace. I guess one option is to create a small C++ program where you initialize Kokkos, and link it to a Kokkos installation with debug enabled, and see if you get the same behavior there.

**Comment by jrmadsen** [comment]
> I think the docs are pretty clear that it can be a bit of a mess with `python setup.py install .. ` or `pip install` + CMake.
> 
> I had to:
> 
> * delete the  `kokkos_compilation` line in cmake/Modules/KokkosPythonKokkos.cmake` because of an error
> * `export Kokkos_DIR=/home/tyler/github_projects/Kokkos/kokkos/build_openmp_cuda`
> * then I run into stuff like: `ninja: error: 'Kokkos::kokkoscore-NOTFOUND', needed by 'kokkos/libpykokkos.cpython-38-x86_64-linux-gnu.so'`
> * then I start messing around with `CMAKE_PREFIX_PATH` to help CMake out a bit...
> 
> Basically, `pykokkos-base` has a hard time finding the the stuff I built following the instructions in this repo. In fact, this repo only mentions `pykokkos-base` with a link to its repo, and actually nothing about it in the install process proper.

example of an external Kokkos install with cuda support + building pykokkos-base: https://github.com/kokkos/pykokkos-base/blob/a195eaaaf0cef83c9b98eb317a4cdb0bb4bcf4f5/.jenkins#L93
